[CodeCompletion] Only type check target statement and related decls in function body

rintaro opened the pull request:
Even after #32086, unrelated statements in closures were type checked.

This PR introduces `typeCheckASTNodeAtLoc()`(renamed from `typeCheckAbstractFunctionBodyAtLoc()`) function that only type checks single body element at a specified location and dependent necessities (i.e. referencing `ValueDecl`s and their dependencies).

* `swift::typeCheckASTNodeAtLoc(DeclContext *, SourceLoc)` type checks single innermost ASTNode at the location. But it leaves any "body" of the node unchecked (i.e. function body, `if` statement body, closure body, etc.)
* Newly implemented on-demand type checking decls:
  * Binding value in control statements (e.g. `ForEachStmt`, `SwitchStmt`, `DoCatchStmt`, etc.)

rdar://problem/63932852

rintaro committed:
[AST] Add a pointer to parent 'TapExpr' expression in 'VarDecl'

Preparation for on-demand 'VarDecl' type checking in interpolated string
literals.

rintaro committed:
[AST] Add a enum for type checking state of ClosureExpr's body

rintaro committed:
[TypeChecker] Invalidate decls in foreach statement

If it's failed to type check the patterns, the sequence expression,
etc., invalidate the pattern and 'VarDecl's in the pattern.

rintaro committed:
[TypeChecker] Factor out StmtChecker::typeCheckASTNode()

For type checking single element in brace statements.

rintaro committed:
[TypeChecker] Consolidate StmtChecker constructors

rintaro committed:
[CodeCompletion] Only type check related statements in function body

Introduce 'TypeCheckSingleASTNode' mode that only type checks single body
element and dependent necessities (i.e. referencing ValueDecls and their
dependencies).

Renamed swift::typeCheckAbstractFunctionBodyAtLoc() to
swift::typeCheckASTNodeAtLoc(DeclContext *, SourceLoc). That type checks
innermost 'ASTNode' at the location. Also, 'TypeCheckSingleASTNode' mode
skips type checking any "body" of the node (i.e. BraceStmt elements for
function body, if statement body, closure body, etc.)

Added on-demand type checking using it:
 - VarDecl in TapExpr
 - ParamDecl in ClosureExpr
 - Return type of ClosureExpr
 - Binding value in control statements
   (e.g. ForEachStmt, SwitchStmt, DoCatchStmt, etc.)

rdar://problem/63932852

rintaro commented:
@swift-ci Please smoke test

rintaro committed:
[Sema] Remove StmtChecker.TargetTypeCheckLoc

The location itself is not used anymore. Use 'TypeCheckSingleASTNode'
instead.

rintaro commented:
@swift-ci Please smoke test

rintaro committed:
[CodeCompletion] Propagate "skipping body" flag per request basis

Instead of using a global type checker flag.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

rintaro commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2384,12 +2404,14 @@ NamingPatternRequest::evaluate(Evaluator &evaluator, VarDecl *VD) const {
+    // Try type checking parent control statement.
     if (auto parentStmt = VD->getParentPatternStmt()) {
-      if (auto LCS = dyn_cast<LabeledConditionalStmt>(parentStmt)) {
-        TypeChecker::typeCheckConditionForStatement(LCS, VD->getDeclContext());
```
Now `typeCheckConditionForStatement()` can be moved from `TypeCheckConstraints.cpp` to `TypeCheckStmt.cpp`, and be internalized in the file. I will do it in a follow up PR.

CodaFi commented on lib/Sema/TypeCheckConstraints.cpp:
```diff
@@ -2237,29 +2240,45 @@ bool TypeChecker::typeCheckPatternBinding(PatternBindingDecl *PBD,
+    // Invalidate the pattern and the var decl.
+    stmt->getPattern()->setType(ErrorType::get(Context));
+    stmt->getPattern()->forEachVariable([&](VarDecl *var) {
+      if (var->hasInterfaceType() && !var->getType()->hasError())
```
```suggestion
      if (var->hasInterfaceType() && !var->isInvalid())
```

Plus a comment explaining that we want to try to preserve the types here if they're not actually invalid.

CodaFi replied:
Come to think of it, you don't really need the `hasInterfaceType` anymore.

rintaro replied:
`hasInterfaceType()` is still needed because `isInvalid()` calls `getInterfaceType()` which calls `typeCheckASTNodeAtLoc()` for this `ForEach` statement.

As for `getType()->hasError()` -> `isInvalid()`, I just copied this from here: https://github.com/apple/swift/blob/4f106019d616ab220f660bbf0559360e1bc1cbe0/lib/Sema/TypeCheckConstraints.cpp#L2320-L2326
Should I fix that too?

rintaro replied:
Done `getType()->hasError()`→ `isInvalid()` for both instances.

CodaFi commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2215,20 +2216,39 @@ InterfaceTypeRequest::evaluate(Evaluator &eval, ValueDecl *D) const {
+    }
+
+    // Try type checking the closure if it hasn't.
+    if (auto *closure = dyn_cast<ClosureExpr>(PD->getDeclContext())) {
```
Whoa, this is a whole lot of work to do in the interface type request. If you need this information, you may as well just typecheck the closure up front.

rintaro replied:
I can remove this block because it turned out that this branch is actually unreachable. `typeCheckASTNodeAtLoc()` type checks surrounding closures anyway.
https://github.com/apple/swift/blob/c4095fc074aba6e9eb6dca26c2d0be770dc4fa71/lib/Sema/TypeCheckStmt.cpp#L1960-L1968

But I don't really understand your argument here. Could you clarify your concern with this on-demand type checking for getting interface types of the decl?

rintaro replied:
Removed this block from the P.

xedin commented on include/swift/AST/Expr.h:
```diff
@@ -3768,7 +3766,25 @@ class SerializedAbstractClosureExpr : public SerializedLocalDeclContext {
+
+    /// The body was typechecked with the enclosing closure.
+    /// i.e. single expression closure or function builder closure.
+    TypeCheckedAtOnce,
```
Nit: probably meant to be `TypeCheckedOnce`.

rintaro replied:
Renamed to `TypeCheckedWithSignature` to clarify the meaning.

rintaro committed:
[Sema] Remove on-demand ParamDecl checking for closures

Because they should have been type checked before the request.

rintaro committed:
[AST] Rename TypeCheckedAtOnce to TypeCheckedWithSignature

To clarify the meaning

CodaFi commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2223,12 +2224,20 @@ InterfaceTypeRequest::evaluate(Evaluator &eval, ValueDecl *D) const {
+      // Type check the ASTNode that contains the parent 'TapExpr' if it
+      // hasn't.
+      if (!parentE->getType())
+        swift::typeCheckASTNodeAtLoc(VD->getDeclContext(), parentE->getLoc());
```
This is subject to the same problems we discussed before. Performing a full type check is not appropriate in the interface type request, and *will* cause circularity problems to be surfaced as in valid entries in Code Completion. Consider a variable binding whose initializer makes reference to the variable itself. Computing the interface type of the declaration doesn't necessarily need to typecheck the initializer binding if there is a user-provided contextual type. Now, it will.

CodaFi replied:
Is this can be handled at the top-level as closures can, that would be great.

rintaro replied:
OK, I'm going to give up on-demand type checking for `TapExpr` for now.

CodaFi commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1279,6 +1279,7 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
 public:
   ASTContext &Ctx;
   SourceFile *SF;
+  bool LeaveFunctionBodyUnchecked;
```
Is there a way to reuse DeclChecker::shouldSkipBodyTypechecking so we don't have to have callers thread this around?

CodaFi replied:
If you wish, you could temporarily set the ASTContext in some kind of "code completion mode" if there isn't one already and read it there.

rintaro replied:
Well, it turned out that `LeaveFunctionBodyUnchecked` is never `true` at least for now. I just removed the flag from `DeclChecker` so this file has no diffs anymore. Although there's still a flag in ConstraintSystem, I think it's better than a "global" flag in ASTContext.

rintaro committed:
[Sema] Stop on-demand type checking for 'VarDecl' in 'TapExpr'

rintaro committed:
[Sema] Remove LeaveFunctionBodyUnchecked flag from DeclChecker

At least for now, ide::typeCheckContextAt() doesn't call
'typeCheckASTNodeAtLoc()' with function decl signature position.
Also, 'getIntefaceType()' on function reference doesn't call decl
checker. So we don't need to propagate it.

rintaro commented:
@swift-ci Please smoke test

rintaro committed:
[Sema] Remove on-demand type checking for return type of ClosureExpr

Closure singatures should be type checked at that point anyway.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
Resolved a conflict
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@CodaFi review ping. Any other comments?

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
Resolved conflicts
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test
